File: python/src/txline/sse.py
"""Server-Sent Events parsing and resilient stream iterators."""

from __future__ import annotations

import asyncio
import contextlib
import json
import time
from collections.abc import AsyncIterator, Callable, Iterator
from dataclasses import dataclass
from typing import Any, Generic, TypeVar

from txline.errors import HttpStatusError
from txline.odds import OddsPayload
from txline.scores import Scores

T = TypeVar("T")


@dataclass(frozen=True, slots=True)
class RawSseEvent:
    id: str | None = None
    event: str | None = None
    data: str = ""
    retry: int | None = None


@dataclass(frozen=True, slots=True)
class SseEvent(Generic[T]):
    id: str | None
    event: str | None
    data: T


@dataclass(frozen=True, slots=True)
class StreamOptions:
    fixture_id: int | None = None
    last_event_id: str | None = None
    initial_backoff: float = 1.0
    max_backoff: float = 30.0


class SseDecoder:
    def __init__(self) -> None:
        self._buffer = ""

    def push(self, chunk: bytes) -> list[RawSseEvent]:
        self._buffer += chunk.decode("utf-8")
        events: list[RawSseEvent] = []
        while True:
            split = _split_sse_block(self._buffer)
            if split is None:
                break
            block, self._buffer = split
            event = parse_sse_block(block)
            if event is not None:
                events.append(event)
        return events

    def finish(self) -> RawSseEvent | None:
        if not self._buffer.strip():
            self._buffer = ""
            return None
        event = parse_sse_block(self._buffer)
        self._buffer = ""
        return event


def parse_sse_block(block: str) -> RawSseEvent | None:
    event_id: str | None = None
    event_type: str | None = None
    retry: int | None = None
    data_lines: list[str] = []
    for raw_line in block.splitlines():
        if not raw_line or raw_line.startswith(":"):
            continue
        if ":" in raw_line:
            field, value = raw_line.split(":", 1)
            if value.startswith(" "):
                value = value[1:]
        else:
            field, value = raw_line, ""
        if field == "id":
            event_id = value
        elif field == "event":
            event_type = value
        elif field == "data":
            data_lines.append(value)
        elif field == "retry":
            with contextlib.suppress(ValueError):
                retry = int(value)
    data = "\n".join(data_lines)
    if event_id is None and event_type is None and not data:
        return None
    return RawSseEvent(id=event_id, event=event_type, data=data, retry=retry)


class AsyncSseStreamClient:
    def __init__(self, client: Any, path: str) -> None:
        self._client = client
        self._path = path

    def stream(self, options: StreamOptions | None = None) -> AsyncIterator[SseEvent[Any]]:
        parser = OddsPayload.from_dict if self._path == "/odds/stream" else Scores.from_dict
        return _async_typed_stream(self._client, self._path, options or StreamOptions(), parser)

    def stream_all(self) -> AsyncIterator[SseEvent[Any]]:
        return self.stream()

    def stream_fixture(self, fixture_id: int) -> AsyncIterator[SseEvent[Any]]:
        return self.stream(StreamOptions(fixture_id=fixture_id))


class SyncSseStreamClient:
    def __init__(self, client: Any, path: str) -> None:
        self._client = client
        self._path = path

    def stream(self, options: StreamOptions | None = None) -> Iterator[SseEvent[Any]]:
        parser = OddsPayload.from_dict if self._path == "/odds/stream" else Scores.from_dict
        return _sync_typed_stream(self._client, self._path, options or StreamOptions(), parser)

    def stream_all(self) -> Iterator[SseEvent[Any]]:
        return self.stream()

    def stream_fixture(self, fixture_id: int) -> Iterator[SseEvent[Any]]:
        return self.stream(StreamOptions(fixture_id=fixture_id))


async def _async_typed_stream(
    client: Any,
    path: str,
    options: StreamOptions,
    parser: Callable[[dict[str, Any]], T],
) -> AsyncIterator[SseEvent[T]]:
    last_event_id = options.last_event_id
    backoff = options.initial_backoff
    while True:
        query = [("fixtureId", str(options.fixture_id))] if options.fixture_id is not None else []
        headers = client.auth_headers(True).to_headers()
        headers["Accept"] = "text/event-stream"
        headers["Cache-Control"] = "no-cache"
        if last_event_id is not None:
            headers["Last-Event-ID"] = last_event_id
        async with client._http.stream(
            "GET", client._api_url(path), params=query, headers=headers
        ) as response:
            if response.status_code in {401, 403}:
                await client._refresh_guest_session_after_failure(client.guest_jwt())
                continue
            if not 200 <= response.status_code <= 299:
                body = await response.aread()
                raise HttpStatusError(response.status_code, body)
            backoff = options.initial_backoff
            decoder = SseDecoder()
            async for chunk in response.aiter_bytes():
                for raw_event in decoder.push(chunk):
                    if raw_event.id is not None:
                        last_event_id = raw_event.id
                    if raw_event.retry is not None:
                        backoff = min(raw_event.retry / 1000, options.max_backoff)
                    typed = _typed_event_from_raw(raw_event, parser)
                    if typed is not None:
                        yield typed
            tail = decoder.finish()
            if tail is not None:
                if tail.id is not None:
                    last_event_id = tail.id
                if tail.retry is not None:
                    backoff = min(tail.retry / 1000, options.max_backoff)
                typed = _typed_event_from_raw(tail, parser)
                if typed is not None:
                    yield typed
        await asyncio.sleep(backoff)
        backoff = min(backoff * 2, options.max_backoff)


def _sync_typed_stream(
    client: Any,
    path: str,
    options: StreamOptions,
    parser: Callable[[dict[str, Any]], T],
) -> Iterator[SseEvent[T]]:
    last_event_id = options.last_event_id
    backoff = options.initial_backoff
    while True:
        query = [("fixtureId", str(options.fixture_id))] if options.fixture_id is not None else []
        headers = client.auth_headers(True).to_headers()
        headers["Accept"] = "text/event-stream"
        headers["Cache-Control"] = "no-cache"
        if last_event_id is not None:
            headers["Last-Event-ID"] = last_event_id
        with client._http.stream(
            "GET", client._api_url(path), params=query, headers=headers
        ) as response:
            if response.status_code in {401, 403}:
                client._refresh_guest_session_after_failure(client.guest_jwt())
                continue
            if not 200 <= response.status_code <= 299:
                raise HttpStatusError(response.status_code, response.read())
            backoff = options.initial_backoff
            decoder = SseDecoder()
            for chunk in response.iter_bytes():
                for raw_event in decoder.push(chunk):
                    if raw_event.id is not None:
                        last_event_id = raw_event.id
                    if raw_event.retry is not None:
                        backoff = min(raw_event.retry / 1000, options.max_backoff)
                    typed = _typed_event_from_raw(raw_event, parser)
                    if typed is not None:
                        yield typed
            tail = decoder.finish()
            if tail is not None:
                if tail.id is not None:
                    last_event_id = tail.id
                if tail.retry is not None:
                    backoff = min(tail.retry / 1000, options.max_backoff)
                typed = _typed_event_from_raw(tail, parser)
                if typed is not None:
                    yield typed
        time.sleep(backoff)
        backoff = min(backoff * 2, options.max_backoff)


def _typed_event_from_raw(
    raw_event: RawSseEvent, parser: Callable[[dict[str, Any]], T]
) -> SseEvent[T] | None:
    if not raw_event.data or (raw_event.event or "").lower() == "heartbeat":
        return None
    data = parser(json.loads(raw_event.data))
    return SseEvent(id=raw_event.id, event=raw_event.event, data=data)


def _split_sse_block(buffer: str) -> tuple[str, str] | None:
    lf = buffer.find("\n\n")
    crlf = buffer.find("\r\n\r\n")
    if lf == -1 and crlf == -1:
        return None
    if crlf != -1 and (lf == -1 or crlf < lf):
        return buffer[:crlf], buffer[crlf + 4 :]
    return buffer[:lf], buffer[lf + 2 :]

File: go/txline/sse.go
package txline

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

type RawSSEEvent struct {
	ID    string
	Event string
	Data  string
	Retry time.Duration
}

type SSEEvent[T any] struct {
	ID    string `json:"id,omitempty"`
	Event string `json:"event,omitempty"`
	Data  T      `json:"data"`
}

type SSEDecoder struct {
	buffer strings.Builder
}

func (d *SSEDecoder) Push(bytes []byte) ([]RawSSEEvent, error) {
	d.buffer.Write(bytes)
	value := d.buffer.String()
	var events []RawSSEEvent
	for {
		block, rest, ok := splitSSEBlock(value)
		if !ok {
			break
		}
		if event, ok := ParseSSEBlock(block); ok {
			events = append(events, event)
		}
		value = rest
	}
	d.buffer.Reset()
	d.buffer.WriteString(value)
	return events, nil
}

func (d *SSEDecoder) Finish() (RawSSEEvent, bool) {
	value := d.buffer.String()
	d.buffer.Reset()
	if strings.TrimSpace(value) == "" {
		return RawSSEEvent{}, false
	}
	return ParseSSEBlock(value)
}

func ParseSSEBlock(block string) (RawSSEEvent, bool) {
	var event RawSSEEvent
	var data strings.Builder
	scanner := bufio.NewScanner(strings.NewReader(block))
	for scanner.Scan() {
		line := scanner.Text()
		line = strings.TrimSuffix(line, "\r")
		if line == "" || strings.HasPrefix(line, ":") {
			continue
		}
		field, value, found := strings.Cut(line, ":")
		if !found {
			value = ""
		} else {
			value = strings.TrimPrefix(value, " ")
		}
		switch field {
		case "id":
			event.ID = value
		case "event":
			event.Event = value
		case "data":
			data.WriteString(value)
			data.WriteByte('\n')
		case "retry":
			if retryMS, err := strconv.ParseInt(value, 10, 64); err == nil && retryMS >= 0 {
				event.Retry = time.Duration(retryMS) * time.Millisecond
			}
		}
	}
	dataValue := data.String()
	event.Data = strings.TrimSuffix(dataValue, "\n")
	if event.ID == "" && event.Event == "" && event.Data == "" && event.Retry == 0 {
		return RawSSEEvent{}, false
	}
	return event, true
}

type StreamOptions struct {
	FixtureID      *int64
	LastEventID    string
	InitialBackoff time.Duration
	MaxBackoff     time.Duration
}

func DefaultStreamOptions() StreamOptions {
	return StreamOptions{
		InitialBackoff: time.Second,
		MaxBackoff:     30 * time.Second,
	}
}

type Stream[T any] struct {
	events <-chan SSEEvent[T]
	errs   <-chan error
}

func (s *Stream[T]) Events() <-chan SSEEvent[T] { return s.events }
func (s *Stream[T]) Errors() <-chan error       { return s.errs }

func (o OddsClient) Stream(ctx context.Context, options StreamOptions) *Stream[OddsPayload] {
	return streamJSON[OddsPayload](ctx, o.client, "/odds/stream", options)
}

func (o OddsClient) StreamAll(ctx context.Context) *Stream[OddsPayload] {
	return o.Stream(ctx, DefaultStreamOptions())
}

func (o OddsClient) StreamFixture(ctx context.Context, fixtureID int64) *Stream[OddsPayload] {
	opts := DefaultStreamOptions()
	opts.FixtureID = &fixtureID
	return o.Stream(ctx, opts)
}

func (s ScoresClient) Stream(ctx context.Context, options StreamOptions) *Stream[Scores] {
	return streamJSON[Scores](ctx, s.client, "/scores/stream", options)
}

func (s ScoresClient) StreamAll(ctx context.Context) *Stream[Scores] {
	return s.Stream(ctx, DefaultStreamOptions())
}

func (s ScoresClient) StreamFixture(ctx context.Context, fixtureID int64) *Stream[Scores] {
	opts := DefaultStreamOptions()
	opts.FixtureID = &fixtureID
	return s.Stream(ctx, opts)
}

func streamJSON[T any](ctx context.Context, client *Client, path string, options StreamOptions) *Stream[T] {
	if options.InitialBackoff <= 0 {
		options.InitialBackoff = time.Second
	}
	if options.MaxBackoff <= 0 {
		options.MaxBackoff = 30 * time.Second
	}
	events := make(chan SSEEvent[T])
	errs := make(chan error, 1)
	go func() {
		defer close(events)
		defer close(errs)
		lastEventID := options.LastEventID
		backoff := options.InitialBackoff
		for {
			if err := ctx.Err(); err != nil {
				return
			}
			retry, err := streamOnce(ctx, client, path, options.FixtureID, &lastEventID, events)
			if err != nil && !errors.Is(err, context.Canceled) && !errors.Is(err, context.DeadlineExceeded) {
				select {
				case errs <- err:
				default:
				}
			}
			if retry > 0 {
				backoff = minDuration(retry, options.MaxBackoff)
			}
			if err := sleepContext(ctx, backoff); err != nil {
				return
			}
			backoff = minDuration(backoff*2, options.MaxBackoff)
		}
	}()
	return &Stream[T]{events: events, errs: errs}
}

func streamOnce[T any](ctx context.Context, client *Client, path string, fixtureID *int64, lastEventID *string, events chan<- SSEEvent[T]) (time.Duration, error) {
	resp, err := client.sseResponse(ctx, path, fixtureID, *lastEventID)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
		stale, _ := client.GuestJWT()
		_, refreshErr := client.refreshGuestAfterFailure(ctx, &stale)
		if refreshErr != nil {
			return 0, refreshErr
		}
		resp.Body.Close()
		resp, err = client.sseResponse(ctx, path, fixtureID, *lastEventID)
		if err != nil {
			return 0, err
		}
		defer resp.Body.Close()
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		return 0, &HTTPStatusError{StatusCode: resp.StatusCode, Body: body}
	}

	decoder := SSEDecoder{}
	buffer := make([]byte, 32*1024)
	var retry time.Duration
	for {
		n, readErr := resp.Body.Read(buffer)
		if n > 0 {
			rawEvents, err := decoder.Push(buffer[:n])
			if err != nil {
				return retry, err
			}
			for _, raw := range rawEvents {
				if raw.ID != "" {
					*lastEventID = raw.ID
				}
				if raw.Retry > 0 {
					retry = raw.Retry
				}
				typed, ok, err := typedSSEEvent[T](raw)
				if err != nil {
					return retry, err
				}
				if !ok {
					continue
				}
				select {
				case events <- typed:
				case <-ctx.Done():
					return retry, ctx.Err()
				}
			}
		}
		if readErr != nil {
			if readErr == io.EOF {
				if raw, ok := decoder.Finish(); ok {
					if raw.ID != "" {
						*lastEventID = raw.ID
					}
					if raw.Retry > 0 {
						retry = raw.Retry
					}
					typed, ok, err := typedSSEEvent[T](raw)
					if err != nil {
						return retry, err
					}
					if ok {
						select {
						case events <- typed:
						case <-ctx.Done():
							return retry, ctx.Err()
						}
					}
				}
				return retry, nil
			}
			return retry, readErr
		}
	}
}

func (c *Client) sseResponse(ctx context.Context, path string, fixtureID *int64, lastEventID string) (*http.Response, error) {
	query := url.Values{}
	if fixtureID != nil {
		query.Set("fixtureId", strconv.FormatInt(*fixtureID, 10))
	}
	reqURL, err := url.Parse(c.apiURL(path, query))
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL.String(), nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "text/event-stream")
	req.Header.Set("Cache-Control", "no-cache")
	if lastEventID != "" {
		req.Header.Set("Last-Event-ID", lastEventID)
	}
	headers, err := c.AuthHeaders(true)
	if err != nil {
		return nil, err
	}
	headers.Apply(req.Header)
	return c.http.Do(req)
}

func typedSSEEvent[T any](raw RawSSEEvent) (SSEEvent[T], bool, error) {
	var zero SSEEvent[T]
	if raw.Event == "heartbeat" || raw.Data == "" {
		return zero, false, nil
	}
	var data T
	if err := json.Unmarshal([]byte(raw.Data), &data); err != nil {
		return zero, false, err
	}
	return SSEEvent[T]{
		ID:    raw.ID,
		Event: raw.Event,
		Data:  data,
	}, true, nil
}

func splitSSEBlock(buffer string) (string, string, bool) {
	lf := strings.Index(buffer, "\n\n")
	crlf := strings.Index(buffer, "\r\n\r\n")
	idx, sepLen := -1, 0
	switch {
	case lf >= 0 && (crlf < 0 || lf < crlf):
		idx, sepLen = lf, 2
	case crlf >= 0:
		idx, sepLen = crlf, 4
	default:
		return "", "", false
	}
	return buffer[:idx], buffer[idx+sepLen:], true
}

func sleepContext(ctx context.Context, d time.Duration) error {
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-timer.C:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

func minDuration(a, b time.Duration) time.Duration {
	if a < b {
		return a
	}
	return b
}
